### PR TITLE
Http retries

### PR DESCRIPTION
**Pull request by DaveCTurner**
Tearing my hair out over this one. Sometimes Chrome just seems not to completely start up, which makes my tests fail with a ResponseTimeout exception. It seems to have got more frequent with a recent Chrome update - almost every test run has at least one session which fails to start up correctly.

This is a sticking-plaster approach - just retry the request if it times out, and only give up if it fails 10 times in a row. PR opened for discussion in case you know of a better way to do this, rather than because I think it's a serious long-term fix.

Cheers,

David

### CODE REVIEW

**Comment by kallisti-dev** [comment]
Anything noteworthy from the Selenium server logs?

**Comment by kallisti-dev** [comment]
I'm going to mess around with some stuff to see if I can identify the cause. My initial guess is that the lower bound dependency for http-client needs to be bumped, based on issue #69 and the relevant fix at https://github.com/snoyberg/http-client/issues/116. While that issue is only reported on Windows, it appears to be similar and there may be other networking problems that will be resolved by enforcing newer versions of http-client. If you could identify which version of http-client you're using that would greatly help my search.

If tweaking the http-client dependency doesn't work, then the next best place to examine would be Selenium commits and issues.

**Comment by DaveCTurner** [comment]
Sure, here's some more details.

It's a `hspec` testsuite that uses `hspec-webdriver` to interface to this library, which then talks (via `http-client`) to a WebDriver server (we have tried 2.43.1, 2.45.0 and 2.46.0). This then talks to a ChromeDriver (we have tried 2.12, 2.15 and 2.16) which then talks to Chrome (currently 44.0.2403.155 but various other versions in the recent past too). All of the above are on the same machine.

The way we've got it set up at the moment involves using a different Chrome session per functional area, which is now in the region of 20. It is increasingly rare that all 20 of these manage to start successfully, causing the test failures we're seeing. There doesn't seem to be a pattern to the failures - it's different each run, and sometimes we've seen two or three fail. The failure is always a ResponseTimeout HttpException and always at session creation. Of course, the number of functional areas has also been increasing so the apparent creeping flakiness of the test suite could well be to do with that.

We've taken a `tcpdump` trace of a full run. Murphy's law applied, of course: the first such run didn't blooming well fail. When we did capture a failure, the effect we saw is a whole sequence of HTTP requests that receive no response: the test suite process POSTs to WebDriver to create a new session, which causes a POST to ChromeDriver which then does a GET on the debugging port of the Chrome instance it's just started. Eventually `http-client` gives up, although the other requests remained stuck until past the end of the run. Some of them were even merrily exchanging TCP keepalives.

I guessed that the upstream requests are blocked on the downstream one, so started looking at what that Chrome instance is playing at. It is indeed running, and a by-hand GET to the debugging port does indeed hang. It's running in `xvfb` so we can't see it directly, but a screenshot of the `xvfb` desktop was blank.

A bit of faffing around with `strace` and `lsof` sort-of hints that it's waiting for one of its child processes to do something, but the child process seems to be stuck in a loop calling `epoll` and immediately receiving `ETIMEDOUT`. I did have a half-hearted look at the Chrome source code but frankly lost the will to continue at around that point so gave up and stuck this retry loop in. The tests have been rock solid since then.

In short, I couldn't find any evidence that the Haskell libs were doing anything wrong - it looks like something weird going on at the other end of the pipeline, in Chrome itself.

A configurable retry count would certainly be nice. If you would like me to attempt a more refined PR along those lines then I can do so. Would you add the retry-count parameter to the `Capabilities` type?

**Comment by kallisti-dev** [comment]
Thanks for the detailed report. I might pass this along to the Selenium dev team as it sounds like something they might be interested in, or might have more information on.

> In short, I couldn't find any evidence that the Haskell libs were doing anything wrong - it looks like something weird going on at the other end of the pipeline, in Chrome itself.

This is often the case in my experience. The implementation of the Haskell library is very straightforward so very few logical issues arise. I suppose much of that stability can be attributed to the nature of Haskell code. Almost all of the maintenance involves interfacing between these much larger dependencies which do the real heavy lifting, and keeping our API up to date with changes in those dependencies.

> The way we've got it set up at the moment involves using a different Chrome session per functional area, which is now in the region of 20. It is increasingly rare that all 20 of these manage to start successfully, causing the test failures we're seeing.

So it sounds possible that ChromeDriver is breaking under the stress. Do you run tests concurrently? Perhaps there's a deadlock or race condition happening somewhere along the pipeline. I don't know for certain the thread-safety of our various dependencies and target servers. For example, I don't know what happens when you send multiple concurrent requests with http-client using different Managers, which would be the default arrangement if you make a `runSession` call for each individual thread.

If concurrency doesn't appear to be an issue, perhaps tweaking the test suite by adding delays or rate limiting would prevent the hang-up.

> A configurable retry count would certainly be nice. If you would like me to attempt a more refined PR along those lines then I can do so. Would you add the retry-count parameter to the Capabilities type?

So far `Capabilities` is only used to implement capabilities as defined within the WebDriver spec, with hs-webdriver-specific configuration being specified within `WDConfig` and then translated into the actual session state stored/access by `WDSession'. This is mostly an artifact of the simplicity of the interface I initially created, being more-or-less a direct implementation of the spec without too much conceptual abstraction. It has the benefit of keeping the implementation very easy to maintain, since there's parity between the location of data within the configuration structs and how they're implemented. While not perfect, In the short term this is the preferred approach for implementing new configuration options.

As more client-side features have been added, the weakness of that approach has become apparent; it's not at all intuitive, I think, for a user of the library to specify session configuration in multiple locations. The solution I have in mind is to hide `Capabilities` from the simplest use cases, and instead use only `WDConfig` as a front-facing configuration interface. The `WDConfig` struct is then used to build `Capabilities` at session creation. A more advanced user could then access `Capabilities` to tweak settings, but it would not be mandatory for setting basic configuration such as browser choice. While not directly related to your problem, any feedback or suggestions on this proposed API would be helpful.

I would recommend keeping the number of retries at 1 by default, since I feel this is a feature that you would only want to use after you've discovered a problem. Retrying HTTP requests by default might hide undiscovered network issues if someone isn't looking carefully enough.

**Comment by DaveCTurner** [comment]
The tests are run sequentially, but there's no pause between the end of one session and the start of the next so it's possible it's something to do with one Chrome not having fully shut down before the next one starts up. They're all started with a different `--user-data-dir` argument so shouldn't be interacting, but maybe they are. Could be a race condition in the Chrome-to-Chrome interaction, failing to handle the exiting instance going away at just the wrong moment, perhaps?

I prefer retrying to pausing between sessions because if we paused we'd have to decide for how long (which may vary depending on other load on the test machine etc. etc). The retrying approach typically works first time, without a pause, and if it doesn't then the HTTP timeout is in effect a large pause.

I've pushed a commit that exposes my magical `10` as a parameter in `WDConfig` as you suggest, with the plumbing via `WDSession`. It preserves the existing behaviour by default by asking for zero retries (i.e. a single try).

**Comment by DaveCTurner** [comment]
Thanks!

Just spotted I left a `hPutStrLn stderr` in this PR, Oops. Suggest either removing it or else adding a `webdriver:` prefix so it's clear to others where that message is coming from.
